libs/utils/trace_analysis.py

```diff
@@ -978,3 +977,265 @@ def getCPUFrequencyResidency(self, cpu):
+        """
+        if not self.trace.hasEvents('cpu_frequency'):
+            logging.warn('Events [cpu_frequency] not found, '\
+                         'plot DISABLED!')
+            return
+        if not self.trace.hasEvents('cpu_idle'):
+            logging.warn('Events [cpu_idle] not found, '\
+                         'plot DISABLED!')
+            return
+
+        if cpus is None:
+            # Generate plots only for available CPUs
+            cpufreq_data = self.trace.df('cpu_frequency')
+            _cpus = range(cpufreq_data.cpu.max()+1)
+        else:
+            _cpus = listify(cpus)
+
+        # Split between big and LITTLE CPUs ordered from higher to lower ID
+        _cpus.reverse()
+        big_cpus = [c for c in _cpus if c in self.platform['clusters']['big']]
+        little_cpus = [c for c in _cpus if c in
+                       self.platform['clusters']['little']]
+        _cpus = big_cpus + little_cpus
+
+        # Precompute active and total time for each CPU
+        residencies = []
+        xmax = 0.0
+        for c in _cpus:
+            r = self.getCPUFrequencyResidency(c)
+            residencies.append(ResidencyData('CPU{}'.format(c), r))
+
+            max_time = r.total.max().values[0]
+            if xmax < max_time:
+                xmax = max_time
+
+        self._generateFrequencyResidencyPlots(residencies, 'cpu', xmax,
+                                              pct, active)
+
+    def plotClusterFrequencyResidency(self, clusters=None,
+                                      pct=False, active=False):
+        """
+        Plot the frequency residency in a given cluster, i.e. the amount of
+        time cluster `cluster` spent at frequency `f_i`. By default, both 'big'
+        and 'LITTLE' clusters data are plotted.
+
+        Requires the following trace events:
+            - cpu_frequency
+            - cpu_idle
+
+        :param clusters: name of the clusters to be plotted (all of them by
+            default)
+        :type clusters: str ot list(str)
+
+        :param pct: plot residencies in percentage
+        :type pct: bool
+
+        :param active: for percentage plot specify whether to plot active or
+            total time. Default is TOTAL time
+        :type active: bool
+        """
+        if not self.trace.hasEvents('cpu_frequency'):
+            logging.warn('Events [cpu_frequency] not found, '\
+                         'plot DISABLED!')
+            return
+        if not self.trace.hasEvents('cpu_idle'):
+            logging.warn('Events [cpu_idle] not found, '\
+                         'plot DISABLED!')
+            return
+
+        # Assumption: all CPUs in a cluster run at the same frequency, i.e. the
+        # frequency is scaled per-cluster not per-CPU. Hence, we can limit the
+        # cluster frequencies data to a single CPU
+        if not self.trace.freq_coherency:
+            logging.warn('Cluster frequency is not coherent! plot DISABLED!')
```

Comment:
For consistency, better use a comma: "Cluster frequency is not coherent, plot DISABLED!"